x-pack/plugins/security_solution/server/endpoint/routes/actions/response_actions.ts

```diff
@@ -290,50 +292,77 @@ function responseActionRequestHandler<T extends EndpointActionDataParameterTypes
   const logger = endpointContext.logFactory.get('responseActionsHandler');
 
   return async (context, req, res) => {
+    logger.debug(`response action [${command}]:\n${dump(req.body)}`);
+
+    // Note:  because our API schemas are defined as module static variables (as opposed to a
+    //        `getter` function), we need to include this additional validation here, since
+    //        `agent_type` is included in the schema independent of the feature flag
+    if (
+      req.body.agent_type &&
+      !endpointContext.experimentalFeatures.responseActionsSentinelOneV1Enabled
+    ) {
+      return errorHandler(
+        logger,
+        res,
+        new CustomHttpRequestError(`[request body.agent_type]: feature is disabled`, 400)
+      );
+    }
+
     const user = endpointContext.service.security?.authc.getCurrentUser(req);
     const esClient = (await context.core).elasticsearch.client.asInternalUser;
     const casesClient = await endpointContext.service.getCasesClient(req);
-    const actionsClient = new EndpointActionsClient({
-      esClient,
-      casesClient,
-      endpointContext,
-      username: user?.username ?? 'unknown',
-    });
+    const connectorActions = (await context.actions).getActionsClient();
+    const responseActionsClient: ResponseActionsClient = getResponseActionsClient(
+      req.body.agent_type || 'endpoint',
+      {
+        esClient,
+        casesClient,
+        endpointService: endpointContext.service,
+        username: user?.username || 'unknown',
+        connectorActions,
+      }
+    );
 
     try {
       let action: ActionDetails;
 
       switch (command) {
         case 'isolate':
-          action = await actionsClient.isolate(req.body);
+          action = await responseActionsClient.isolate(req.body);
           break;
 
         case 'unisolate':
-          action = await actionsClient.release(req.body);
+          action = await responseActionsClient.release(req.body);
           break;
 
         case 'running-processes':
-          action = await actionsClient.runningProcesses(req.body);
+          action = await responseActionsClient.runningProcesses(req.body);
           break;
 
         case 'execute':
-          action = await actionsClient.execute(req.body as ExecuteActionRequestBody);
+          action = await responseActionsClient.execute(req.body as ExecuteActionRequestBody);
```

Comment:
Instead of casting maybe we could pass the type down from caller?

Reply:
The issue is that you don't know the exact type in `req.body` since this handler is setup to process all response actions. So here, the `case` is almost like a type guard, but unfortunately it would more complicated code to actually implement with type guards.

In these instances here, its fairly safe to cast because we are checking the `command` and we know for sure _(because it passed schema validation)_ that if that check matches, the body of the request is in fact the type we are casting to.